// @proton/wrap-constants/src/parsers/currency.ts
import * as z from 'zod'

export interface CurrencyStat {
    supply    : string;
    max_supply: string;
    issuer    : string;
}
export interface CurrencyStats {
   [key: string]: CurrencyStat;
}

export const rateParser = z.object({
    counterCurrency   : z.string(),
    price             : z.number(),
    priceChangePercent: z.number(),
    marketCap         : z.number(),
    volume            : z.number(),
    timestamp         : z.number()
})
export type Rate = z.infer<typeof rateParser>

export const exchangeRateParser = z.object({
    contract: z.string(),
    symbol  : z.string(),
    rank    : z.number(),
    rates   : z.array(rateParser)
})
export type ExchangeRate = z.infer<typeof exchangeRateParser>

export const exchangeRatesParser = z.array(exchangeRateParser)
export type ExchangeRates = z.infer<typeof exchangeRatesParser>

export const bloksRatesParser = z.array(z.object({
    contract: z.string(),
    symbol  : z.string(),
    rank    : z.number(),
    rates   : z.array(rateParser)
}))

export type BloksRate = {
    key: string;
    symbol: string;
    account: string;
    chain: string;
    supply: {
        circulating: number;
        max: number;
        precision: number;
        last_update: string;
    };
    metadata: {
        name: string;
        website: string;
        logo: string;
        created_at: string;
        desc: string;
    };
    pairs: [
        {
            id: string;
            pair_base: string;
            pair_quote: string;
            exchange: string;
            quote:{
                price: number
                volume_base_24h: number
                volume_quote_24h: number
                market_cap: number
                price_usd: number
                volume_usd_24h: number
            },
            percentage_daily_volume: number
        },
    ];
    price: {
        quotes: {
            [key: string]: number
        },
        usd: number;
        volume_base_24h: number;
        volume_usd_24h: number;
        marketcap_usd: number;
        change_24hr: number;
    },
    rank: number;
    chain_rank: number;
}

export type BloksRates = BloksRate[]
